rust/src/server.rs

```diff
@@ -22,6 +22,18 @@ const CHAT_INGEST_RECEIVE_BUFFER: usize = 10;
 const BROADCAST_BUFFER: usize = 32;
 const REQUEST_TIMEOUT: Duration = Duration::from_secs(1);
 
+const X_AUTH_TAG: &str = "x-auth-tag";
+
+fn extract_auth_tag<T>(req: &Request<T>) -> RpcResult<String> {
+    Ok(req
+        .metadata()
+        .get(X_AUTH_TAG)
+        .ok_or_else(|| Status::internal("missing auth tag"))?
+        .to_str()
+        .map_err(|_| Status::internal("failed to decode auth tag"))?
```

Comment:
Can you include the reason here as well?

Reply:
Yeah, errors throughout should probably have more context than they do now.